sdks/python/apache_beam/dataframe/transforms.py

```diff
@@ -0,0 +1,244 @@
+      """
+      def __init__(self, inputs, is_grouping):
+        self.inputs = set(inputs)
+        self.is_grouping = is_grouping or len(self.inputs) > 1
+        self.ops = []
+        self.outputs = set()
+
+    # First define some helper functions.
+    def output_is_partitioned_by_index(expr, stage):
+      if expr in stage.inputs:
+        return stage.is_grouping
+      elif expr.preserves_partition_by_index():
+        if expr.requires_partition_by_index():
+          return True
+        else:
+          return all(
+              output_is_partitioned_by_index(arg, stage) for arg in expr.args())
+      else:
+        return False
+
+    def _partition_by_index(df, levels=None, parts=10):
+      if levels is None:
+        levels = list(range(df.index.nlevels))
+      elif isinstance(levels, (int, str)):
+        levels = [levels]
+      hashes = sum(
+          pd.util.hash_array(df.index.get_level_values(level))
+          for level in levels)
+      for k in range(parts):
+        yield k, df[hashes % parts == k]
+
+    def common_stages(stage_lists):
+      if stage_lists:
+        for stage in stage_lists[0]:
+          if all(stage in other for other in stage_lists[1:]):
+            yield stage
+
+    @memoize
+    def expr_to_stages(expr):
+      assert expr not in inputs
+      # First attempt to compute this expression as part of an existing stage,
+      # if possible.
+      for stage in common_stages([expr_to_stages(arg) for arg in expr.args()
+                                  if arg not in inputs]):
+        if (not expr.requires_partition_by_index() or
+            all(output_is_partitioned_by_index(arg, stage)
+                for arg in expr.args())):
+          break
+      else:
+        # Otherwise, compute this expression as part of a new stage.
+        stage = Stage(expr.args(), expr.requires_partition_by_index())
+        for arg in expr.args():
+          if arg not in inputs:
+            expr_to_stages(arg).append(stage)
+            expr_to_stage(arg).outputs.add(arg)
+      stage.ops.append(expr)
+      # This is a list as given expression may be available in many stages.
+      return [stage]
+
+    def expr_to_stage(expr):
+      # Any will do; the first requires the fewest intermediate stages.
+      return expr_to_stages(expr)[0]
+
+    # Ensure each output is computed.
+    for expr in outputs.values():
+      if expr not in inputs:
+        expr_to_stage(expr).outputs.add(expr)
+
+    @memoize
+    def stage_to_result(stage):
+      return {expr._id: expr_to_pcoll(expr)
+              for expr in stage.inputs} | ComputeStage(stage)
+
+    @memoize
+    def expr_to_pcoll(expr):
+      if expr in inputs:
+        return inputs[expr]
+      else:
+        return stage_to_result(expr_to_stage(expr))[expr._id]
+
+    # Now we can compute and return the result.
+    return {k: expr_to_pcoll(expr) for k, expr in outputs.items()}
+
+
+def memoize(f):
```

Comment:
We already have these defined in a couple of places (I see one in `bundle_processor.py` and one in `apache_beam.transforms.external`). It seems silly to redefine it everywhere, but the alternative (pulling it into some util file) also seems silly. I guess the best option is just replace them all with `functools.lru_cache` after we drop python 2.